### ISSUE TITLE
insert_primary_key_rows does not return primary keys of inserted rows.

### ISSUE THREAD

**Issue by mathbeal** [sqlalchemy/sqlalchemy#6194]
**Describe the bug**

After multiple insertions at same time, `insert_primary_key_rows` return `[(None,), (None,)]`. The primary key of each inserted row is expected. 

The code works as expected with on insertion. `insert_primary_key` has the pk of the inserted row. but it does work with multiples insertions.

**Expected behavior**

`[(1,), (2,)]` expected as it's an in memory database just created, and it's the first insertions.

**To Reproduce**
```py
from sqlalchemy import MetaData
from sqlalchemy import create_engine
from sqlalchemy import Table, Column, Integer, String
from sqlalchemy import insert, select

engine = create_engine("sqlite+pysqlite:///:memory:", echo=True, future=True)
metadata = MetaData()

user_table = Table(
     "user_account",
     metadata,
     Column('id', Integer, primary_key=True),
     Column('name', String(30)),
     Column('fullname', String)
)
metadata.create_all(engine)

# Insertion of 2 rows at same time.
with engine.connect() as conn:
    result = conn.execute(
        insert(user_table),
        [
            {"name": "sandy", "fullname": "sandy test"},
            {"name": "tata", "fullname": "tata test"},
        ]
    )
    conn.commit()
    print(result.inserted_primary_key_rows)

# Insertion of 1 row
with engine.connect() as conn:
    result = conn.execute(
        insert(user_table),
        [
            {"name": "ouba", "fullname": "ouba test"},
        ]
    )
    conn.commit()
    print(result.inserted_primary_key_rows)


with engine.connect() as conn:
    result = conn.execute(select(user_table))
    for row in result:
        print(row)

```

**Error**

```
2021-04-03 21:53:50,895 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-04-03 21:53:50,895 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("user_account")
2021-04-03 21:53:50,895 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("user_account")
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine 
CREATE TABLE user_account (
	id INTEGER NOT NULL, 
	name VARCHAR(30), 
	fullname VARCHAR, 
	PRIMARY KEY (id)
)


2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine [no key 0.00013s] ()
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine COMMIT
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine INSERT INTO user_account (name, fullname) VALUES (?, ?)
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine [generated in 0.00024s] (('sandy', 'sandy test'), ('tata', 'tata test'))
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine COMMIT
[(None,), (None,)]
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine INSERT INTO user_account (name, fullname) VALUES (?, ?)
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine [generated in 0.00020s] ('ouba', 'ouba test')
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine COMMIT
[(3,)]
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine SELECT user_account.id, user_account.name, user_account.fullname 
FROM user_account
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine [generated in 0.00023s] ()
(1, 'sandy', 'sandy test')
(2, 'tata', 'tata test')
(3, 'ouba', 'ouba test')
2021-04-03 21:53:50,908 INFO sqlalchemy.engine.Engine ROLLBACK
```

**Versions.**
 - OS: Windows-10-10.0.19041-SP0
 - Python: Python 3.8.8 (tags/v3.8.8:024d805, Feb 19 2021, 13:18:16) [MSC v.1928 64 bit (AMD64)] on win32
 - SQLAlchemy: 1.4.5
 - Database: sqlite+pysqlite:///:memory
 - DBAPI: sqlite3 2.6.0

**Have a nice day!**

**Comment by zzzeek** [comment]
The accessor you're using I would say right now is completely not documented as it doesn't actually do what you are trying to make it do for the SQLite backend.

Except in one extremely specific (and not at the moment documented) case, there's no way to get back inserted primary key values when you use DBAPI executemany(), which is what happens when you run an insert with a list of parameters as you are doing above.    

the accessor there is explicitly to support the approach used by psycopg2's [psycopg2 fast execution helpers](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#psycopg2-fast-execution-helpers), which SQLAlchemy has special support for.    There is a plan to generalize these execution helpers to all of the PostgreSQL backends in #6047 .

The approach used by the fast execution helper is that DBAPI .execute(), not executemany(), is used, and the INSERT statement is rewritten in the form:

   INSERT INTO table (x, y) VALUES (v1, v2), (v3, v4), (v5, v6), .... RETURNING <pk1>

I was about to say that the above is not possible with SQLite because they dont support RETURNING, however as they add features all the time I just checked and it looks like [RETURNING support was just added three weeks ago](https://www.sqlite.org/releaselog/3_35_0.html).    So I've added #6195  to look into this but it remains to be seen if it even works with the pysqlite driver (probably does).     SQLAlchemy would need to add support for this and current Python releases won't have a bundled SQLite that new in any case unless you build manually.

So to sum up:

1. there's no generalized way to get inserted primary keys back from multiple rows except by using a specially formed INSERT statement
2. SQLAlchemy uses this specially formed INSERT statement transparently within the conn.execute() method for exactly the psycopg2 driver only right now, soon for the other postgresql drivers
3. There's not a lot of other DBs that support multi VALUES + multi-row RETURNING, likely newer mariadb, I have to check if SQL server supports the multi VALUES syntax with INSERT
4. SQLite looks to support both the multi VALUES and RETURNING syntax but this is brand new and we don't support it yet.

**Comment by zzzeek** [comment]
ive rewritten the docs for the current situation which you can see as follows:

https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=inserted_primary_key_rows#sqlalchemy.engine.BaseCursorResult.inserted_primary_key_rows


maybe in a few months we can have this happening for more backends.
